pkg/bundle/bundle.go

```diff
@@ -0,0 +1,267 @@
+		}
+
+		if !filepath.IsAbs(deref) {
+			deref = filepath.Join(
+				filepath.Dir(path),
+				deref,
+			)
+		}
+
+		info, err := os.Stat(deref)
+		return deref, info, err
+	}
+
+	var patterns []gitignore.Pattern
+	if file, err := os.Open(filepath.Join(directory, shpIgnoreFilename)); err == nil {
+		defer file.Close()
+
+		domain := split(directory)
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := strings.TrimSpace(scanner.Text())
+			if len(line) != 0 && !strings.HasPrefix(line, "#") {
+				patterns = append(patterns, gitignore.ParsePattern(line, domain))
+			}
+		}
+
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+	}
+
+	matcher := gitignore.NewMatcher(patterns)
+
+	var buf bytes.Buffer
+	var tw = tar.NewWriter(&buf)
+	defer tw.Close()
+
+	err := filepath.WalkDir(directory, func(path string, d fs.DirEntry, err error) error {
+		// Bail out on path errors
+		if err != nil {
+			return err
+		}
+
+		// Skip files on the ignore list
+		if matcher.Match(split(path), d.IsDir()) {
+			if d.IsDir() {
+				return filepath.SkipDir
+			}
+
+			return nil
+		}
+
+		info, err := d.Info()
+		if err != nil {
+			return err
+		}
+
+		header, err := tar.FileInfoHeader(info, path)
+		if err != nil {
+			return err
+		}
+
+		header.Name, err = filepath.Rel(directory, path)
+		if err != nil {
+			return err
+		}
+
+		switch {
+		case info.Mode().IsDir():
+			return tw.WriteHeader(header)
+
+		case info.Mode().IsRegular():
+			if err := tw.WriteHeader(header); err != nil {
+				return err
+			}
+
+			return write(tw, path)
+
+		case info.Mode()&os.ModeSymlink == os.ModeSymlink:
+			deref, info, err := followSymLink(path)
+			if err != nil {
+				return err
+			}
+
+			header, err = tar.FileInfoHeader(info, deref)
+			if err != nil {
+				return err
+			}
+
+			header.Name, err = filepath.Rel(directory, path)
+			if err != nil {
+				return err
+			}
+
+			if err := tw.WriteHeader(header); err != nil {
+				return err
+			}
+
+			return write(tw, deref)
+
+		default:
+			return fmt.Errorf("unsupported file type: %s", path)
+		}
+	})
+
+	return &buf, err
+}
+
+// Unpack reads a tar stream and writes the content into the local file system
+// with all files and directories.
+func Unpack(in io.Reader, targetPath string) error {
+	var tr = tar.NewReader(in)
+	for {
+		header, err := tr.Next()
+		switch {
+		case err == io.EOF:
+			return nil
+
+		case err != nil:
+			return err
+
+		case header == nil:
+			continue
+		}
+
+		var target = filepath.Join(targetPath, header.Name)
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if err := os.MkdirAll(target, os.FileMode(header.Mode)); err != nil {
+				return err
+			}
+
+		case tar.TypeReg:
+			// Edge case in which that tarball did not have a directory entry
+			dir, _ := filepath.Split(target)
+			if err := os.MkdirAll(dir, os.FileMode(0755)); err != nil {
+				return err
+			}
+
+			file, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
```

Comment:
I won't hold the PR up for this, but will cite that per our previous conversation, that I think the `os.OpenFile` through the `io.Copy` should be in a separate method, so you can employ `defer file.Close` in that separate method, and not get the golang complaint about using defer's directly in a for loop.